### ISSUE TITLE
torch.nn.utils.clip_grad_norm_: bad GPU utilization due to GPU-data-dependent control-flow

### ISSUE THREAD

**Issue by mautier** [pytorch/pytorch#60691]
## Issue description

During CUDA training, using [`torch.nn.utils.clip_grad_norm_`](https://github.com/pytorch/pytorch/blob/d69c22dd61a2f006dcfe1e3ea8468a3ecaf931aa/torch/nn/utils/clip_grad.py#L9) negatively affects my GPU's utilization.

Specifically, the implementation contains 3 instances where the CPU control-flow depends on the result of GPU computations, forcing stream synchronizations:
1. `if total_norm.isnan() or total_norm.isinf():` ([link](https://github.com/pytorch/pytorch/blob/d69c22dd61a2f006dcfe1e3ea8468a3ecaf931aa/torch/nn/utils/clip_grad.py#L43)):
    - The result of `isnan` is copied back to host pageable memory, to apply standard boolean short-circuit logic.
    - If no short-circuit occurs (the standard finite case), the same is done for the result of `isinf`.
3. `if clip_coef < 1:` ([link](https://github.com/pytorch/pytorch/blob/d69c22dd61a2f006dcfe1e3ea8468a3ecaf931aa/torch/nn/utils/clip_grad.py#L57)):
    - The comparison is run on the device as it should be.
    - The final result is copied back to host pageable memory to pick a branch.

Since gradient clipping is typically performed before calling `Optimizer.step()`, the result is that the optimizer's GPU computations (or whatever is run after clipping) suffer from poor pipelining and thus poor GPU utilization.

## Code example

The following is a dummy but complete model training example using gradient norm clipping:
```python
import torch


def main():
    cpu = torch.device("cpu")
    cuda = torch.device("cuda:0")

    tensorboard_profile_dir = "/tmp/tensorboard_profile/"
    with torch.profiler.profile(
        activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA],
        schedule=torch.profiler.schedule(wait=0, warmup=2, active=1, repeat=1),
        with_stack=True,
        on_trace_ready=torch.profiler.tensorboard_trace_handler(tensorboard_profile_dir),
    ) as profile:

        batch_size = 1
        num_feat = 1500  # Increase this to make the cuda sync in clip_grad_norm_ more dramatic.
        model = torch.nn.Linear(
            in_features=num_feat, out_features=num_feat, bias=False, device=cuda
        )

        optimizer = torch.optim.SGD(model.parameters(), lr=1e-3)

        with torch.profiler.record_function("init data"):
            # Random input data, and fixed regression target.
            x = torch.randn(batch_size, num_feat, device=cuda)
            target = torch.full((num_feat,), 100.0, device=cuda)

        for _ in range(3):
            with torch.profiler.record_function("forward + loss"):
                y = model(x)
                l1_loss = (y - target).abs().mean()

            with torch.profiler.record_function("backward"):
                optimizer.zero_grad()
                l1_loss.backward()

            with torch.profiler.record_function("clip grad"):
                torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=0.1, norm_type="inf")

            optimizer.step()

            profile.step()

    print("Saved profile to", tensorboard_profile_dir)


if __name__ == "__main__":
    main()
```

Corresponding profile in tensorboard:
![clip_grad_norm_sync_issue annotated](https://user-images.githubusercontent.com/5213905/123330344-c2c81a00-d4f2-11eb-9efd-d35cccc6df4b.png)

## Workaround

In my training code, to work around the issue I changed the implementation of `clip_grad_norm_`:
1. I removed the `isnan()` and `isinf()` checks (and crossed my toes that my gradient stays well-behaved).
2. I replaced the `if clip_coef < 1:` by an unconditional scaling:
  ```diff
  -if clip_coef < 1:
  -    [... multiply gradients by clip_coef ...]
  +clip_coef = torch.clamp(clip_coef, max=1.0)
  +[... multiply gradients by clip_coef ...]
  ```

Doing so removed all 3 control-flow dependencies, and gave me a nice boost to GPU utilization (85% -> 95%) and training speed.
Obviously doing 1. is a pretty ill-advised; maybe there's some mechanism to defer raising the exception / printing the warning till later, when the result of the computation is available?
2. on the other hand merely adds no-op computations in the case where no clipping is necessary (but makes performance more consistent I suppose!).

<details>

<summary>Tensorboard trace with the workarounds</summary>

![image](https://user-images.githubusercontent.com/5213905/123333063-23a52180-d4f6-11eb-960c-2aa7cebf07e0.png)

</details>

## System Info

- PyTorch or Caffe: PyTorch
- How you installed PyTorch (conda, pip, source): docker, using the `pytorch/pytorch:1.9.0-cuda11.1-cudnn8-devel` image.
- OS: Linux, Ubuntu 20.04
- PyTorch version: 1.9.0
- Python version: 3.7.10
- CUDA/cuDNN version: NVIDIA Driver 460.80, CUDA 11.2, cuDNN 8
- GPU models and configuration: GeForce MX150

cc @ngimel @VitalyFedyunin @heitorschueroff

**Comment by ngimel** [comment]
Inf and nan checks are added by #53843, you can disable them by calling clip_grad_norm with `error_if_nonfinite=False` https://pytorch.org/docs/master/generated/torch.nn.utils.clip_grad_norm_.html#torch.nn.utils.clip_grad_norm_. Unfortunately, there's no way to raise a warning until you do a sync (which will likely be after optimizer step, at which point your parameters will already be corrupted)

**Comment by mautier** [comment]
Thanks for the quick response!

What about the `if clip_coef < 1`? If that conditional remains as it is, then the function syncs anyway, which makes the nan/inf syncs benign.

**Comment by ngimel** [comment]
It should be possible to get rid of `clip_coef` synchronization either like you suggest, or creating some custom op that would skip pointless multiplication.

**Comment by jbschlosser** [comment]
Thanks @ngimel, I agree we should move the `isnan()` / `isinf()` checks behind the `error_if_nonfinite` flag (requires keeping it opt-in rather than making it default True in the future, which is fine). `clip_coef` sync should be removed as well by usage of `clamp()`.

We'd accept a PR implementing these changes, if you're up for it @mautier.

**Comment by mautier** [comment]
Sounds good, I'll put up a PR shortly then.
